packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/widgets/toolbar.ts

```diff
@@ -0,0 +1,109 @@
+import { notify } from '@affine/component';
+import {
+  generateUrl,
+  type UseSharingUrl,
+} from '@affine/core/hooks/affine/use-share-url';
+import { getAffineCloudBaseUrl } from '@affine/core/modules/cloud/services/fetch';
+import { EditorService } from '@affine/core/modules/editor';
+import { I18n } from '@affine/i18n';
+import type { MenuItemGroup } from '@blocksuite/affine-components/toolbar';
+import type { MenuContext } from '@blocksuite/blocks';
+import { LinkIcon } from '@blocksuite/icons/lit';
+import type { FrameworkProvider } from '@toeverything/infra';
+
+export function createToolbarMoreMenuConfig(framework: FrameworkProvider) {
+  return {
+    configure: <T extends MenuContext>(groups: MenuItemGroup<T>[]) => {
```

Comment:
This API is a litter bit hack and not universally useful. I prefer APIs like:

Reply:
This `configure` API needs to know the contents of `groups` to be overwritten, which TypeScript cannot infer.
```
configure: <T extends MenuContext>(groups: MenuItemGroup<T>[]) => MenuItemGroup<T>[]
```

Reply:
Here are some grouping definitions, and the type definitions in this area will be strengthened later.

https://github.com/toeverything/blocksuite/blob/445b1943814b9a38b40acd57a79feb29283cb9cf/packages/blocks/src/root-block/widgets/element-toolbar/more-menu/button.ts#L19-L26

There are also horizontal actions, which will take some time to sort out.

Reply:
Yeah, abstraction is necessary. You can create a refactoring plan and complete it later step by step.

Reply:
Good catch.

I also recommend that all goups are controlled by AFFiNE.
In blocksuite, just read a complete configuration.
But the current migration cost is relatively high. Take baby steps first

Reply:
Yep, eventually the rendering of toolbar components only depends on config in different block specs, flexible and scalable.